pallets/tinlake-investor-pool/src/lib.rs

```diff
@@ -560,13 +614,79 @@ pub mod pallet {
 				.collect()
 		}
 
-		fn is_epoch_valid(
-			_reserve: T::Balance,
-			_target: &[(T::Balance, T::Balance)],
-			_solution: &[(Perquintill, Perquintill)],
-		) -> bool {
-			// TODO: Implement this
-			true
+		pub fn is_epoch_valid(
+			reserve: T::Balance,
+			max_reserve: T::Balance,
+			target: &[(T::Balance, T::Balance)],
+			tranche_ratios: &[Perquintill],
+			current_tranche_values: &[T::Balance],
+		) -> DispatchResult {
+			let acc_supply: T::Balance = target
+				.iter()
+				.fold(Zero::zero(), |sum: T::Balance, tranche| {
+					sum.checked_add(&tranche.0).unwrap()
```

Comment:
This (and redeem below it) should be multiplying the target `Balance` with the solution `Perquintill` to get the actual Balance amounts for this epoch solution. Otherwise we're just checking the targets over and over with each solution we pass.

Reply:
I was accounting on the targets to take care of having that baked in already. But yeah, we either do it within this function or by the invoker. I am fine either way, so I will consume here the solution ratios.